packages/api/amplify_api/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart

```diff
@@ -0,0 +1,425 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'dart:async';
+
+import 'package:amplify_api/src/graphql/web_socket/blocs/ws_subscriptions_bloc.dart';
+import 'package:amplify_api/src/graphql/web_socket/services/web_socket_service.dart';
+import 'package:amplify_api/src/graphql/web_socket/state/web_socket_state.dart';
+import 'package:amplify_api/src/graphql/web_socket/state/ws_subscriptions_state.dart';
+import 'package:amplify_api/src/graphql/web_socket/types/web_socket_types.dart';
+import 'package:amplify_api/src/graphql/web_socket/types/ws_subscriptions_event.dart';
+import 'package:amplify_flutter/amplify_flutter.dart';
+import 'package:async/async.dart';
+
+part '../types/web_socket_event.dart';
+
+/// {@template amplify_api.web_socket_bloc}
+/// Internal state machine for the web socket connections.
+/// Listens for [WebSocketEvent] and maps them to appropriate business logic.
+/// {@endtemplate}
+class WebSocketBloc with AWSDebuggable, AmplifyLoggerMixin {
+  /// {@macro api.web_socket_bloc}
+  WebSocketBloc({
+    required AWSApiConfig config,
+    required AmplifyAuthProviderRepository authProviderRepo,
+    required WebSocketService wsService,
+  }) {
+    _config = config;
+    _authProviderRepo = authProviderRepo;
+    final subBlocs = <String, WsSubscriptionBloc<Object?>>{};
+
+    _currentState = DisconnectedState(
+      _config,
+      _authProviderRepo,
+      NetworkState.connected,
+      IntendedState.connected,
+      wsService,
+      subBlocs,
+    );
+    final blocStream = _wsEventStream.asyncExpand(_eventTransformer);
+    _subscription = blocStream.listen(_emit);
+  }
+
+  // Config and auth repo together determine how to authorize connection URLs
+  // and subscription registration messages.
+  late AmplifyAuthProviderRepository _authProviderRepo;
```

Comment:
question: Why are these `late` here when they were already assigned on line 39/40 above? Could this part be removed?

Reply:
nm, I think I get this.